### PR TITLE
Add genotype upload for raredisease

### PR DESCRIPTION
**Pull request by rannick**
### Added

-

### Changed

-

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/meta/upload/gt.py [hunk 1]
```diff
@@ -39,97 +43,139 @@ def data(self, analysis: Analysis) -> dict:
...
+            samples_sex[sample_id] = {
+                "pedigree": case_sample.sample.sex,
+                "analysis": self.get_analysis_sex_raredisease(qc_metrics_file, sample_id),
+            }
+        return samples_sex
 
-    @staticmethod
-    def is_suitable_for_genotype_upload(case_obj: Case) -> bool:
-        """Check if a cancer case is contains WGS and normal sample."""
+    def get_analysis_sex_raredisease(self, qc_metrics_file: Path, sample_id: Sample) -> str:
```
Comment (ivadym) [inline cg/meta/upload/gt.py:155]:
```suggestion
    def get_analysis_sex_raredisease(self, qc_metrics_file: Path, sample_id: str) -> str:
```

File: cg/meta/upload/gt.py [hunk 2]
```diff
@@ -39,97 +43,139 @@ def data(self, analysis: Analysis) -> dict:
...
+        """Return analysis sex for each sample of an analysis."""
+        qc_metrics: list[MetricsBase] = self.get_parsed_qc_metrics_data_raredisease(qc_metrics_file)
+        return str(
+            next(
+                metric.value
+                for metric in qc_metrics
+                if metric.name == RAREDISEASE_PREDICTED_SEX_METRIC and metric.id == sample_id
+            )
+        )
```
Comment (ivadym) [inline cg/meta/upload/gt.py:164]: Do we need to use `next` for a specific reason here? 

```suggestion
for metric in qc_metrics:
        if metric.name == RAREDISEASE_PREDICTED_SEX_METRIC and metric.id == sample_id:
            return str(metric.value)
```

File: cg/meta/upload/gt.py [hunk 3]
```diff
@@ -39,97 +43,139 @@ def data(self, analysis: Analysis) -> dict:
...
+        self, case: Case, hk_version: Version
+    ) -> dict[str, dict[str, str]]:
+        """Return sex information from StatusDB and from analysis prediction (stored Housekeeper QC metrics file)."""
+        qc_metrics_file: Path = self.get_qcmetrics_file(hk_version)
+        samples_sex: dict[str, dict[str, str]] = {}
+        for case_sample in case.links:
+            sample_id: str = case_sample.sample.internal_id
+            samples_sex[sample_id] = {
+                "pedigree": case_sample.sample.sex,
+                "analysis": self.get_analysis_sex_raredisease(qc_metrics_file, sample_id),
```
Comment (ivadym) [inline cg/meta/upload/gt.py:151]: names args

File: cg/meta/upload/gt.py [hunk 4]
```diff
@@ -39,97 +43,139 @@ def data(self, analysis: Analysis) -> dict:
...
+        """Parse and return a QC metrics file."""
+        qcmetrics_raw: dict = ReadFile.get_content_from_file(
+            file_format=FileFormat.YAML, file_path=qc_metrics
         )
+        return [MetricsBase(**metric) for metric in qcmetrics_raw["metrics"]]
+
+    def get_qcmetrics_file(self, hk_version_obj: Version) -> Path:
+        """Return a QC metrics file path."""
+        hk_qcmetrics: Path = self.hk.files(
+            version=hk_version_obj.id, tags=HkAnalysisMetricsTag.QC_METRICS
```
Comment (ivadym) [inline cg/meta/upload/gt.py:176]: It accepts a set of metrics

```suggestion
            version=hk_version_obj.id, tags={HkAnalysisMetricsTag.QC_METRICS}
```

File: cg/meta/upload/gt.py [hunk 5]
```diff
@@ -39,97 +43,139 @@ def data(self, analysis: Analysis) -> dict:
...
+        ).first()
+        if hk_qcmetrics is None:
+            raise FileNotFoundError("QC metrics file not found for the given hk version.")
```
Comment (ivadym) [inline cg/meta/upload/gt.py:179]:
```suggestion
            raise FileNotFoundError("QC metrics file not found for the given Housekeeper version.")
```

File: cg/meta/upload/gt.py [hunk 6]
```diff
@@ -39,97 +43,139 @@ def data(self, analysis: Analysis) -> dict:
...
+    def get_parsed_qc_metrics_data_raredisease(self, qc_metrics: Path) -> list[MetricsBase]:
+        """Parse and return a QC metrics file."""
+        qcmetrics_raw: dict = ReadFile.get_content_from_file(
+            file_format=FileFormat.YAML, file_path=qc_metrics
         )
+        return [MetricsBase(**metric) for metric in qcmetrics_raw["metrics"]]
+
+    def get_qcmetrics_file(self, hk_version_obj: Version) -> Path:
+        """Return a QC metrics file path."""
+        hk_qcmetrics: Path = self.hk.files(
```
Comment (ivadym) [inline cg/meta/upload/gt.py:175]: Files doesn't return a `Path` or a `File`, but a `Query` object. Maybe what you want to use here is `get_latest_file` instead?

File: cg/meta/upload/gt.py [hunk 7]
```diff
@@ -39,97 +43,139 @@ def data(self, analysis: Analysis) -> dict:
...
+        samples_sex: dict[str, dict[str, str]] = {}
+        for case_sample in case.links:
+            if case_sample.sample.is_tumour:
+                continue
+            sample_id: str = case_sample.sample.internal_id
             samples_sex[sample_id] = {
-                "pedigree": link_obj.sample.sex,
-                "analysis": analysis_sexes[sample_id],
+                "pedigree": case_sample.sample.sex,
+                "analysis": Sex.UNKNOWN,
```
[the remark this answers is not in the document]
Comment (rannick) [reply cg/meta/upload/gt.py:109]: the pedigree as Mapped[str], the analysis as str?

[the remark this answers is not in the document]
Comment (ivadym) [reply cg/meta/upload/gt.py:109]: I was wondering if `Sex.UNKNOWN` will automatically be converted to a string within that dictionary as it does within an f-string. I guess we'll find out if it fails the upload. Just something to keep in mind if we notice any weird behavior

**Comment by rannick** [comment]
Closing as test building went too far

**Pull request closed without merging**